File: Baekjun/24.9.3/3066. 브리징 시그널/solution.py
T = int(input())
from bisect import bisect_left

while T:
    T -= 1
    N = int(input())

    dp = []
    for _ in range(N):
        n = int(input())
        idx = bisect_left(dp, n)

        if idx == len(dp):
            dp.append(n)
        else:
            dp[idx] = n

    print(len(dp))

File: Baekjun/24.9.3/3066. 브리징 시그널/solution.js
/** @format */

const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().split("\n");

let T = Number(input[0]);
let line = 0;

const bisect_left = (n, arr) => {
  let [l, r] = [-1, arr.length];
  while (r > l + 1) {
    const mid = Math.floor((l + r) / 2);

    if (arr[mid] >= n) {
      r = mid;
    } else {
      l = mid;
    }
  }
  return r;
};

while (T) {
  let dp = [];
  T -= 1;

  const N = Number(input[++line]);
  for (let i = 0; i < N; i++) {
    const n = Number(input[++line]);
    const idx = bisect_left(n, dp);

    if (idx === dp.length) {
      dp.push(n);
    } else {
      dp[idx] = n;
    }
  }
  console.log(dp.length);
}
